Added a new function that stages a dataset, also integrated with the download function

lucascottolavina opened the pull request:
A new function has been created, in the file `admix/manager.py`, named `bring_online(did,rse)`. It requests, through gfal commands, data to be pre-staged (after checking that they are stored in a tape-based storage system) and waits until data are all cached on the disk buffer (hence, available to be more efficiently copied or downloaded). This PR includes:

- the implementation of this new function
- the modification of the `admix/downloader.py` function to that it can optionally (`stage=True` as input) activate a pre-stage before a download
- the new flag `--stage` in `bin/admix-download` executable, that, if added, enables this feature before the actual download.

As a result, for users interested to download data from tape-based storage systems, it suffices to add the flag `--stage` when they run the admix-download command. 

In the case this flag is activated for data not stored in tape-based systems (so far a not elegant solution is done, hardcoding the names of the tape-based RSEs), the program will just skip the staging. So, a priori, the stage flag could also be activated in any occasion if you don't know where data are stored. This new feature still needs to be tested massively, that's why it can be currently activated only with that flag. Once it is fully validated, we can make the stage active by default, so that the choice of staging will be transparent to users.

lucascottolavina committed:
Added a new function that stages a dataset, also integrated with the download function

FaroutYLq commented on admix/downloader.py:
```diff
@@ -78,7 +79,7 @@ def download_dids(dids, num_threads=8, **kwargs):
 
 def download(did, chunks=None, location='.',  tries=3, metadata=True,
-             only_metadata=False, num_threads=5, rse=None):
+             only_metadata=False, num_threads=8, rse=None, stage=None):
```
Is there a motivation for bumping the download thread number to 8?

lucascottolavina replied:
not really, it was probably due to many tests I have done. I will have it back to 5

FaroutYLq commented on admix/manager.py:
```diff
@@ -397,3 +400,65 @@ def clean_local_dir(path, before_straxen_version, ensure_rucio=False, dry_run=Tr
+def bring_online(did,rse):
+    print("Bringing online {0} from {1}".format(did,rse))
+
+    if rse not in ['SURFSARA_USERDISK','CNAF_TAPE3_USERDISK','CNAF_TAPE2_USERDISK','CNAF_TAPE_USERDISK','CCIN2P3_USERDISK']:
```
Better to put something like
```python
TAPE_RSES = ['SURFSARA_USERDISK','CNAF_TAPE3_USERDISK','CNAF_TAPE2_USERDISK','CNAF_TAPE_USERDISK','CCIN2P3_USERDISK']
```
in top of the script for better maintenance of hardcoded configs.

FaroutYLq replied:
Also what about?
```
| NIKHEF2_USERDISK    | 421.270 TB | 400.000 TB | 0.000 B      |
| NIKHEF_USERDISK     | 711.558 GB | 275.000 TB | 274.288 TB   |
```
I forgot if they are mixed or tapes...

lucascottolavina replied:
I will make the change.
NIKHEF and NIKHEF2 are 100% disk (but, careful, NIKHEF should never be used, the numbers there are just what rucio has in the catalogue but the physical storage has been dismissed)

FaroutYLq reviewed (commented):
Thanks a lot @lucascottolavina for this important PR. If I understand it correctly, this if we need to stage, the admix-download command may stay there for days right? If so I think if it makes sense to add a warning message when staging, to recommend users to put such command into a screen session to avoid "having been waited for days and it dies"

lucascottolavina commented:
Hi @FaroutYLq thanks for your prompt feedback. Certainly it will take quite some time. With the tests done so far, the staging of single datasets (composed by hundreds of files) is relatively efficient. The delay ranges between few minutes to half an hour, depending on many elements (I guess how busy is the tape systems at that moment). I can certainly add a comment, it does not hurt. Then, we can discuss at the telecon, how we can tell people to not arbitrarily launch massive download commands at once.

FaroutYLq reviewed (approved):
Thanks. Looks to me then.

lucascottolavina committed:
Added a timeout to protect from hanging, improved verbosity

lucascottolavina commented:
Hi @FaroutYLq and @dachengx, all changes are done. Basically:
- created a TAPE_RSES outside the function on top of the script (this will be further improved in the future as soon as I have time to modify admix in such a way that it accepts configuration parameters in .xenon_config (this is important to merge the master one day)
- added several protections and warnings: a timeout to avoid infinite waiting for stage (default is one hour); if timeout is set >10 minutes, a warning appears suggesting using a screen session; finally a warning, in case of timeout, suggesting that a possible download could not be as performing as expected since not all files are fully staged
- fixed that change on the number of threads, now it is 5 as it was before

FaroutYLq commented:
Thanks! Please feel free to merge
